### ISSUE TITLE
Outlier identification (and database checking in general)

### ISSUE THREAD

**Issue by tresoldi** [lingpy/lingpy#326]
As suggested by @LinguList , we might think of integrating some basic database checking into lingpy. The very first checks to perform would be looking for duplicate entries (when not allowed), empty entries (when not allowed), and "outliers" both in forms and segments (based in context to state transition probability, as discussed at https://github.com/glottobank/lexibank-data/issues/344 and with an example, a stand-alone script, at https://gist.github.com/tresoldi/2b7f1e0a8a2804f305309b9c12d790f3 )

A tentative command call would be 

```
$ lingpy check dataset
```

and a tentative example would be

```
$ lingpy check bowernpny
[INFO] Entries with lowest probability for column 'Form':
('BOORRI-MUL', -11.868724322304748)
('MURDALANGINY', -11.863348562588262)
('KUDADJININY', -11.757519486334322)
('Gürrügümá', -11.749938834060362)
('NGAPAPPI', -11.648591623587155)
('dʒɤnɤgɾɪ', -11.478116814363315)
('ŋwiyá-ŋa1yuru-badá2 BREAD', -11.471807512938062)
('vδaywvnv', -11.437570303552498)
('BIRRID', -11.366652964636804)
('dʊɹʊmuʊndɪɹɪ', -11.309727075114662)
('nɤndɤ', -11.297346394450063)
('see ‘camp’', -11.217972267114687)
('not listed in other dialects', -11.194399904620557)
('ᶁeɲe:ɲ', -11.183360831037625)
('oŋkiːrdɐl', -11.181269189433603)
('NADJA', -11.140401781783192)
('əruoδ', -11.140401781783192)
('BAALUN', -11.059560993034141)
('ɪndʒɪnjɪ', -10.983952572102384)
('only specific', -10.966429060416962)
```

**Comment by xrotwang** [comment]
If the checks do not require lingpy functionality, they may better go into
pycldf?

Am 22.11.2017 22:34 schrieb "Tiago Tresoldi" <notifications@github.com>:

**Comment by LinguList** [comment]
Yes, however, what I was thinking of is how to organize a workflow of testing whether a dataset is amenable for automatic or clean manual analysis, and here, this involves several steps (and I'm still learning which ones are needed), which are at times dependent but often also independent:

1 a general search for strange stuff (outliner identification)
2 a check for the segmentability of the forms (so that we can create an initial orthography profile)
3 a check for morpheme segmentation (if a form has a word "a b + + b a", this needs to be replaced, as it would imply a morpheme "")
4 simple detection of the major "splitters" (e.g., comma, slash, double-backslash [!]) for multiple word forms
5 synonymy 
6 what are gap characters (we could guess, but people use all kinds of things)
7 checking of the orthography profile (@chrzyki made a first test for orthography profiles)

etc.

I am asking myself: what is the best way to proceed, ideally also so that people less experience with a computer could carry out those tasks. And here, I think it is best to have one package where the checks are carried out instead of telling users that they should use package X for task A, package Y for task B, etc. 

The problem is: tasks clearly belong to different packages. pyclts for example should not be a dependency for lingpy, unless we ship the major segmentation algorithm of lingpy to pyclts, to have sequences and sound classes in one place. pycldf reads and writes cldf files, but cleaning a dataset may also be something we want to do when doing automatic analysis, etc.

Maybe, we should think of making a compatibility package later, something that draws together cli-commands for various tasks, then calling different code from lingpy, outliers, pycldf, pyclts?

**Comment by xrotwang** [comment]
From the perspective of CLDF, I see this as follows:

1. Most of these tasks/checks relate to particular columns in the FormTable - representing successive stages of annotation or refinement. So when we want to attach functionality to these, they should be documented in the Ontology.
2. Once we relate functionality to data via the Ontology, we only have two options: Either the package implementing the functionality requires pycldf to read/write the data, or pycldf provides a plugin mechanism to include functionality from other packages. 

I'm leaning towards the latter, because I think it's better for the end user to become accustomed to the `cldf` command being the main entry point to interacting with CLDF data.
In terms of usability, this would still be acceptable, I think. It would just require installing two python packages instead of just pycldf, provided there's a way to auto-discover plugins. (I think entry points in `setup.py` would be appropriate here.)

So the general idea is: 
- Packages register functionality relating to particular CLDF objects (tables or properties specified via Ontology URLs).
- pycldf looks up the registry and is in charge to dispatch to other packages, feeding them the appropriate data.

**Comment by LinguList** [comment]
sounds good to me. This would also allow, for example, to make a first orthoprofile via CLDF using the current command in lingpy (which would not be easy to be broken off the code, at least not now)?

And in general, I agree: it should be pycldf/cldf as the center, as this is a format specification and other packages work with it.

**Comment by tresoldi** [comment]
I agree, as CLDF integration can also lead to a better outlier identification given the expected easiness in dataset test writing (nobody is considering a general, multipurpose, database curation tool).

This behavior can be mimicked with plain csv code, of course, but information from other fields is helpful: for example, after a first pass for identifying outliers, in which we collect all forms from all languages (useful for finding, say, English glosses), a second round testing form groups by language_id can be helpful in pointing to graphemes that are less probable in a particular language. In fact, it helped catch an error on `bowernpny`.
